Remove team super, rename invincible to super

KebsCS opened the pull request:
Fixes https://github.com/ddnet/ddnet/issues/9117
Helps progress [128 player support]( https://github.com/ddnet/ddnet/pull/10581)
Fixes https://github.com/ddnet/ddnet/issues/4799

Draft, only server-side now. Also renamed `NUM_DDRACE_TEAMS` -> `NUM_ALL_TEAMS` and `TEAM_SUPER`  -> `NUM_JOINABLE_TEAMS` to see the changes more clearly.

Server builds fine, and works correctly with the latest master client. Client won't build now because of the renamed variables.
I'll revert the variable names or rename them to something better and fix the build if this PR gets the green light.


**How this PR works:**
`/super` will now send both invincible and super flags. Their logic was the same in most cases, but also sending the invincible flag adds the sparkles particles. 

Prediction is working perfectly fine. That is done by using `CharacterCore` super and invincible variables (now `IsSuper()`) and removing `Team == TEAM_SUPER` checks. (We can keep the checks on client side to predict servers that have super) 

So far from my testing, the only issue is that doors are incorrectly rendered. But that problem is on the client-side prediction incorrectly removing lasers, even though we receive snaps for them. (wrong super detection)

What should be done on the client-side? 

## Checklist

- [x] Tested the change ingame
- [ ] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test (especially base/) or added coverage to integration test
- [ ] Considered possible null pointers and out of bounds array indexing
- [ ] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

KebsCS committed:
Remove team super, rename invincible to super

ChillerDragon commented:
Yesterday you mentioned a network flag that lets the server tell the client if 64 is the magic super team or a regular team. Are you still planning to add that?

KebsCS commented:
> Could you please leave the name `NUM_DDRACE_TEAMS` it was good. Anything that is related to ddrace teams should contain ddrace in the name to not be confused with regular teams like TEAM_RED and TEAM_SPECTATORS

What do I change `TEAM_SUPER' to? Or leave it because it exists on client side? Or do I remove it's support from client side too? 



> Yesterday you mentioned a network flag that lets the server tell the client if 64 is the magic super team or a regular team. Are you still planning to add that?

This is not needed currently, we still have super but the server just won't use the special team. This can be added in 128 players pr if we want to support more than 64 teams

ChillerDragon commented:
> What do I change `TEAM_SUPER' to? Or leave it because it exists on client side? Or do I remove it's support from client side too?

I would keep the client side support for now. We can still remove it later. But at the very least we should give servers time to update.

KebsCS commented:
Brought back `NUM_DDRACE_TEAMS`.

What should I do to `TEAM_SUPER`? Leave it for client-side, but try to not use it server-side?

ChillerDragon commented:
> Leave it for client-side, but try to not use it server-side?

That is what I would propose yes.

KebsCS committed:
Adapt super on client side

ChillerDragon commented on src/game/server/gamecontext.cpp:
```diff
@@ -3927,9 +3920,7 @@ void CGameContext::RegisterDDRaceCommands()
 	Console()->Register("ninja", "", CFGFLAG_SERVER | CMDFLAG_TEST, ConNinja, this, "Makes you a ninja");
 	Console()->Register("unninja", "", CFGFLAG_SERVER | CMDFLAG_TEST, ConUnNinja, this, "Removes ninja from you");
-	Console()->Register("super", "", CFGFLAG_SERVER | CMDFLAG_TEST, ConSuper, this, "Makes you super");
-	Console()->Register("unsuper", "", CFGFLAG_SERVER, ConUnSuper, this, "Removes super from you");
```
Why did you remove unsuper?

KebsCS replied:
@ChillerDragon so for chat commands this should also be 2 commands, or 1 acting as toggle? Super didnt have chat command before

ChillerDragon replied:
I don't have any opinion on chat commands. Maybe someone else does.

ChillerDragon replied:
There was some discussion about keeping the rcon and chat commands in sync somewhere

ChillerDragon replied:
I have no opinion here. Works for me either way. @heinrich5991 is our specialist for breaking users configs.
@fokkonaut and @Pioooooo were involved in the linked discussions about practice commands and command naming conventions.

Someone should chip in here.

KebsCS replied:
Imo could stay as it was before with super and unsuper

ChillerDragon commented on src/game/server/entities/character.h:
```diff
@@ -274,8 +272,7 @@ class CCharacter : public CEntity
 
-	bool IsSuper() const { return m_Core.m_Super; }
-
+	bool IsSuper() const { return m_Core.m_Super && m_Core.m_Invincible; }
```
Why did this change? What does m_Super = true and m_Invincible = false mean?

ChillerDragon replied:
So we could also just delete m_Invincible?

ChillerDragon replied:
Okay so one is just visual and the other is super prediction I guess.

If that is the case I wish the variable name would communicate that more clearly. Should be called m_Sparkly if it does not actually make you invincible.

KebsCS replied:
So do I remove invincible prediction? And just keep it as sparky for now? I guess removing the prediction isn't big deal, it'll only be bad when non-ddnet server runs an old version with that command

ChillerDragon replied:
Not sure yet. I just want to avoid ending up with two variables that do the same thing. If invincible is only for the visual effect it should be renamed and no prediction should be attached to it if we already have a m_Super property for that. But I can already see Heinrich disliking that he does not want to give visual effects descriptive names (that's why confetti is also not called confetti).

But I would have to read the full code and the prs that added m_Invincible first before being able to make a proper decision here.

ChillerDragon commented:
Yea we have to be really sure nothing breaks here. This pr will probably take some time to get merged. I think we can already split up the `IsValidTeamNumber` into a new pr which can be merged faster. It is a unrelated change and good either ways if we keep super or not.

heinrich5991 commented:
> Some old super features are missing, but can be added in followup pr if needed.

Is that something you plan to do? Or is it just "anything can be added since you can open PRs for anything"? I don't like that formulation, please clarify.

"I'll create a followup PR for old super features that are missing."

"I'm not adding old super features (and I'm not planning on creating followup PRs)."

sjrc6 commented:
> > Some old super features are missing, but can be added in followup pr if needed.
> 
> Is that something you plan to do? Or is it just "anything can be added since you can open PRs for anything"? I don't like that formulation, please clarify.

I'll do the features that aren't related to switches after this is merged if kebs won't.

ChillerDragon commented:
This changes the rcon `super` command to show the sparkle effect which on master is not the case. I doubt anyone has a problem with it. I just wanted to mention it explicitly.

ChillerDragon commented:
On master activating super disables switches visually and also gameplay wise.

With this pr switches get hidden visually but still collide with the player. That is a bit odd. I guess we can move it to a follow up pr but it will look like a bug.


## invincible on master

<img width="258" height="166" alt="image" src="https://github.com/user-attachments/assets/29842bf1-5325-41a7-af44-b6674a4d151a" />

## super on master

<img width="258" height="166" alt="image" src="https://github.com/user-attachments/assets/bffd707f-4053-47cd-a227-9c2a5869ab76" />

## "super" with this pr

<img width="258" height="166" alt="image" src="https://github.com/user-attachments/assets/4f95a5e8-f67e-4489-b657-daad92c63561" />

ChillerDragon commented:
Added the config breaking change because the `super` now toggles and has new arguments. And more importantly because `unsuper` was removed. I think it is fine to break users configs here without deprecation.

Zwelf commented:
> Looks like zwelf is afk, so I am making the call for him. @KebsCS could you please bump the teehistorian version because super physics changed.

Yes, I'm mostly afk right now. Only looking into github mail from time to time. I currently don't have the capacity to do more often. Thanks for making the call. (From glancing at the PR I agree that teehistorian bump would be good.)

ChillerDragon commented on src/game/server/ddracechat.cpp:
```diff
@@ -1335,7 +1337,7 @@ void CGameContext::ConTeam0Mode(IConsole::IResult *pResult, void *pUserData)
 	bool Mode = pController->Teams().TeamFlock(Team);
 
-	if(Team <= TEAM_FLOCK || Team >= TEAM_SUPER)
+	if(!pController->Teams().IsValidTeamNumber(Team))
```
This changed behavior! The error used to kick in for TEAM_FLOCK and now you say only non valid teams cause an error that is without TEAM_FLOCK.

So when running the command /team0mode on master in team 0 (flock) you get:

```
chat/server: *** This team can't have the mode changed
```

and with this change you get:

```
chat/server: *** 'ChillerDragon' enabled team 0 mode. This will make your team behave like team 0.
```

KebsCS replied:
Fixed in https://github.com/ddnet/ddnet/pull/11353

ChillerDragon commented on src/game/server/gamecontext.cpp:
```diff
@@ -1086,7 +1081,7 @@ void CGameContext::OnPreTickTeehistorian()
 			m_TeeHistorian.RecordPlayerTeam(i, 0);
 	}
-	for(int i = 0; i < TEAM_SUPER; i++)
+	for(int i = 0; i < NUM_DDRACE_TEAMS; i++)
```
This changed the loop max from 64 to 65. Is that ok? I guess so because we use the `i` only in `RecordTeamPractice` which uses it to index `CTeam m_aPrevTeams[MAX_CLIENTS];` and `MAX_CLIENTS` is `128` so its at least not a out of bounds.

ChillerDragon commented on src/game/server/teams.cpp:
```diff
@@ -235,7 +234,7 @@ void CGameTeams::Tick()
 	}
 
-	for(int i = 0; i < TEAM_SUPER; i++)
+	for(int i = 0; i < NUM_DDRACE_TEAMS; i++)
```
This also does one more iteration now. Also should be fine. We only use `i` to index.

`int m_aTeamUnfinishableKillTick[NUM_DDRACE_TEAMS];`

AssassinTee replied:
TeamId

ChillerDragon commented on src/game/server/teams.cpp:
```diff
@@ -1357,15 +1341,15 @@ ETeamState CGameTeams::GetTeamState(int Team) const
 bool CGameTeams::TeamLocked(int Team) const
 {
-	if(Team <= TEAM_FLOCK || Team >= TEAM_SUPER)
+	if(!IsValidTeamNumber(Team))
```
This changes behavior. Not sure yet if it is a problem need to give it a closer look.

This used to return false here for TEAM_FLOCK, with this change it no longer does that.

ChillerDragon reviewed (changes requested):
I tested a lot in game today. And gave the entire diff a first review run. I still need to look a bit closer one more time. But I already found one thing that should be fixed. My review also got some comments about me thinking loud. @KebsCS feel free to resolve all the comments that you think need no further attention.

KebsCS commented on src/game/server/teams.cpp:
```diff
@@ -1357,15 +1341,15 @@ ETeamState CGameTeams::GetTeamState(int Team) const
 bool CGameTeams::TeamFlock(int Team) const
 {
-	if(Team <= TEAM_FLOCK || Team >= TEAM_SUPER)
+	if(!IsValidTeamNumber(Team))
```
I'll add the case for `TEAM_FLOCK` back. Not sure why is `TEAM_FLOCK` handled so much differently from `m_aTeamFlock` though

KebsCS commented:
> We should not break anyones local setup that has a simple super bind to test maps

They could just add `/practice` to the bind. Or setup `sv_practice_by_default`, or add it to the bind since it sets practice in t0 too

ChillerDragon commented:
@KebsCS could you separate super from other ddrace teams fully please? So there is no way for them to see each other or interact.

Fixing the weirdness of https://github.com/ddnet/ddnet/pull/11232#issuecomment-3591064782

sjrc6 commented:
> Was interacting with other ddrace teams as super a bug or feature? 

It was a feature, admins for some reason wanted to be able to troll other teams. I don't think any mods use it for anything useful.

KebsCS commented:
> @KebsCS I would really like to progress here. I can offer you to take over. Or do you plan to finish this in the near future?

I'm a bit busy coding other stuff atm. You can take it over, just add me to commit as co-author 🙂

ChillerDragon committed:
Visually contain super tees in their own ddrace team

ChillerDragon committed:
Physically contain super tees in their own ddrace team

https://github.com/ddnet/ddnet/labels/fix-changes-physics

ChillerDragon commented:
Okay I fixed https://github.com/ddnet/ddnet/pull/11232#issuecomment-3591064782. Now super can only interact with its own ddrace team. So players from other teams can not break the super users hook or show up as faded out tees. Super can interact with all players from team 0 if the super user is also in team 0. This cleaned up the code and also removed the troll feature of ddrace teams leaking into each other.

I spent a lot of time closely reviewing and testing kebs changes. Since I now did the rebase and the final fix I guess I can no longer merge this my self. But as far as I know this is ready for merge. @AssassinTee send help.

AssassinTee commented on src/game/client/components/items.cpp:
```diff
@@ -468,7 +468,6 @@ void CItems::OnRender()
 	if(Client()->State() != IClient::STATE_ONLINE && Client()->State() != IClient::STATE_DEMOPLAYBACK)
 		return;
 
-	bool IsSuper = GameClient()->IsLocalCharSuper();
```
Can you also remove IsLocalCharSuper then? It's now unused

ChillerDragon committed:
Remove `IsLocalCharSuper()`

ChillerDragon committed:
Longer variable names

ChillerDragon committed:
Restore old `super` and `unsuper` rcon commands

CC https://github.com/ddnet/ddnet/pull/11232/changes#r2531965644

ChillerDragon commented:
@AssassinTee should be ready now.

ChillerDragon commented on src/game/teamscore.cpp:
```diff
@@ -28,14 +28,14 @@ void CTeamsCore::Team(int ClientId, int Team)
 bool CTeamsCore::CanKeepHook(int ClientId1, int ClientId2) const
 {
-	if(m_aTeam[ClientId1] == (m_IsDDRace16 ? VANILLA_TEAM_SUPER : TEAM_SUPER) || m_aTeam[ClientId2] == (m_IsDDRace16 ? VANILLA_TEAM_SUPER : TEAM_SUPER) || ClientId1 == ClientId2)
+	if(ClientId1 == ClientId2)
```
You want to fully remove `VANILLA_TEAM_SUPER`? not sure about that. I feel like someone will complain if we just break ancient ddrace clients. Its still used in `bool CCharacterCore::IsSwitchActiveCb(int Number, void *pUser)`

I would like to remove it too. But I feel like getting approval for that will slow down the pr merge.

ChillerDragon replied:
Yes you are right. Removed it. Thanks for the idea.

ChillerDragon committed:
Fully remove `VANILLA_TEAM_SUPER`

This prediction and physics edge case is no longer needed
because super can not interact with other teams anymore.

https://github.com/ddnet/ddnet/pull/11232#discussion_r2641160260

ChillerDragon committed:
Use NUM_DDRACE_TEAMS as max instead of TEAM_SUPER

Preperation for https://github.com/ddnet/ddnet/pull/11232#discussion_r2936620112
Which is a part of removing super.
Which is a preperation for 128 slot support.

ChillerDragon committed:
Fully remove TEAM_SUPER also from the client side

https://github.com/ddnet/ddnet/pull/11232#discussion_r2936620112

KebsCS commented on src/game/server/teams.cpp:
```diff
@@ -1350,9 +1334,9 @@ void CGameTeams::ResetSavedTeam(int ClientId, int Team)
-	for(int i = 1; i < TEAM_SUPER; i++)
-		if(m_aTeamState[i] == ETeamState::EMPTY)
-			return i;
+	for(int TeamId = 1; TeamId < NUM_DDRACE_TEAMS - 1; TeamId++) // no TEAM_SUPER
```
`TEAM_SUPER` doesn't exist so comment is outdated.
Also, shouldn't it be without `- 1`?

KebsCS commented:
Closing this PR because I've lost interest in the project due to toxicity towards contributors, hostility from admins towards new features (https://github.com/ddnet/ddnet/pull/9917#issuecomment-3710341303, https://github.com/ddnet/ddnet/pull/11260#issuecomment-3566034471, https://github.com/ddnet/ddnet/pull/11550#discussion_r2692164652), and especially poor decision-making while ignoring the community (https://github.com/ddnet/ddnet/pull/11957)

If someone would like to continue this PR, feel free to do so
